src/Data/BitVector/Sized/Internal.hs

```diff
@@ -1,419 +1,748 @@
 
-import Data.Bits
-import Data.Ix
-import Data.Parameterized
+import Data.BitVector.Sized.Panic (panic)
+
+import qualified Data.Bits as B
+import qualified Numeric as N
+
+import Data.Char (intToDigit)
+import Data.Int
+import Data.Maybe (fromJust)
+import Data.Word
+import Data.Parameterized ( NatRepr
+                          , natValue
+                          , intValue
+                          , widthVal
+                          , addNat
+                          , ShowF
+                          , EqF(..)
+                          , Hashable(..)
+                          , HashableF(..))
+import qualified Data.Parameterized.NatRepr as P
+import qualified Data.Parameterized.Pair as P
 import GHC.Generics
 import GHC.TypeLits
-import Numeric
-import System.Random
-import Text.PrettyPrint.HughesPJClass
-import Text.Printf
-import Unsafe.Coerce (unsafeCoerce)
+import Language.Haskell.TH.Lift (Lift)
+import Numeric.Natural
+import Prelude hiding (abs, or, and, negate, concat, signum)
+import qualified Prelude as Prelude
+
+----------------------------------------
+-- Utility functions
+
+-- | Check that a 'NatRepr' is representable as an 'Int'.
+checkNatRepr :: NatRepr w -> a -> a
+checkNatRepr = checkNatural . natValue
+
+-- | Check that a 'Natural' is representable as an 'Int'.
+checkNatural :: Natural -> a -> a
+checkNatural n a = if n > fromIntegral (maxBound :: Int)
+  then panic "Data.BitVector.Sized.Internal.checkNatural"
+       [show n <> " not representable as Int"]
+  else a
 
 ----------------------------------------
 -- BitVector data type definitions
 
--- | BitVector datatype, parameterized by width.
-data BitVector (w :: Nat) :: * where
-  BV :: NatRepr w -> Integer -> BitVector w
-  deriving Generic
-
--- | 'BitVector' can be treated as a constructor for pattern matching, but to build
--- one you must use the smart constructor `bitVector`.
-pattern BitVector :: NatRepr w -> Integer -> BitVector w
-pattern BitVector wRepr x <- BV wRepr x
-{-# COMPLETE BitVector #-}
-
--- | Construct a bit vector with a particular width, where the width is inferrable
--- from the type context. The input (an unbounded data type, hence with an
--- infinite-width bit representation), whether positive or negative, is silently
--- truncated to fit into the number of bits demanded by the return type.
+-- | Bitvector datatype, parameterized by width.
+data BV (w :: Nat) :: * where
+  -- | We store the value as an 'Integer' rather than a 'Natural',
+  -- since many of the operations on bitvectors rely on a two's
+  -- complement representation. However, an invariant on the value is
+  -- that it must always be positive.
+  BV :: Integer -> BV w
+  deriving (Generic, Show, Read, Eq, Ord, Lift)
+
+instance ShowF BV
+
+instance EqF BV where
+  BV bv `eqF` BV bv' = bv == bv'
+
+instance Hashable (BV w) where
+  hashWithSalt salt (BV i) = hashWithSalt salt i
+
+instance HashableF BV where
+  hashWithSaltF salt (BV i) = hashWithSalt salt i
```

Comment:
Same question as with `EqF`

Reply:
Yeah, I think this is a legitimate gripe. Without a `NatRepr`, I don't see how we can define this instance correctly.

Reply:
If the instance is needed for something, I don't think it is a huge problem, though I suspect it isn't because you would need `testEquality` to use it (I'd think) and you can't write that either.

Reply:
I agree - I don't think the instance could every really be useful.